### ISSUE TITLE
[QUESTION] About loss curve

### ISSUE THREAD

**Issue by YangLinzhuo** [thunlp/Seq1F1B#5]
您好，我使用代码在单机 8 卡上训练了 2.5b 配置的 GPT 模型，序列长度是 32k，序列切分数为 2，发现前 100 步的 loss 曲线会有一定的误差，具体入下图所示

<img width="623" alt="Image" src="https://github.com/user-attachments/assets/e8ee2c49-50b5-4479-89c0-7863920df559" />

loss 的误差大概在百分位。

按照论文里面的说法是对 loss 曲线没有影响的

![Image](https://github.com/user-attachments/assets/78f9cc08-0448-47e8-ae6d-ee947b5ea90c)

我具体有两个问题：
1. 请问这个对 loss 没有影响的判定标准大概是怎样的呢？
2. 请问试验用的具体环境配置是怎样的呢？比如使用的 nvidia 镜像版本，flash_attn 的版本这些

谢谢！

**Comment by MayDomine** [comment]
1. Mathematically, it is equivalent to the original loss calculation. However, if a `loss_mask` exists, the `report_loss` will differ. This is because the way `report_loss` is calculated for each micro-batch is as follows:
 `report_loss = (loss * loss_mask).sum() / loss_mask.sum()`
 Here, `loss_mask` is used to indicate whether each token participates in the loss calculation. When the micro-batch is split into two sequences and calculated separately, it becomes:
 `report_loss = (loss_chunk_1 * loss_mask_1).sum() / loss_mask_1.sum() + (loss_chunk_2 * loss_mask_2).sum() / loss_mask_2.sum()`
 From this, it can be observed that when there are zeros in the `loss_mask`, the `report_loss` calculated in segments is not mathematically equivalent to the original method. However, this only affects the `report_loss`, and the model's backpropagation using `loss` remains unaffected. Therefore, observing different loss values may not necessarily impact the model's performance.
 The second factor affecting the loss comes from low-precision accumulation. In the original single-batch backpropagation, gradient accumulation is not required. However, when the sequence is split into two segments for separate backpropagation, an additional accumulation step is needed. Due to the inherent errors in low-precision types like bf16/fp16 during accumulation, this can also lead to differences in the training loss compared to the original method. The same issue applies to all methods that involve gradient accumulation. When gradients are accumulated across multiple micro-batches or steps, the use of low-precision data types (such as bf16 or fp16) can introduce numerical errors during the accumulation process.
2. NVIDIA image: nvcr.io/nvidia/pytorch:24.07-py3  
flash_attn version: 2.5.6

**Comment by YangLinzhuo** [comment]
感谢您的回复！还想请问一下，论文里面关于对 loss 没有影响的说明，是通过实验对比过多步 loss 曲线得到的？还是直接根据数学的等价性判断的呢？

**Comment by MayDomine** [comment]
> 感谢您的回复！还想请问一下，论文里面关于对 loss 没有影响的说明，是通过实验对比过多步 loss 曲线得到的？还是直接根据数学的等价性判断的呢？

Since this issue is mathematically obvious, we did not include experimental results in the article.
After completing the initial implementation, we verified the impact of sequence segmentation on loss based on both simple data parallelism and pipeline parallelism (e.g. Seq1F1B). An error of 1%-5% exists, but it does not affect the model's performance.

**Comment by YangLinzhuo** [comment]
好的，非常感谢您的解答！
